### ISSUE TITLE
CPU usage increased dramatically 0.8.1-RC1 -> master

### ISSUE THREAD

**Issue by bobrik** [mesosphere/marathon#1497]
I built and deployed 964e43035830bc48c56e4853e39ed44553383427. Running against 0.22.1 masters:

![image](https://cloud.githubusercontent.com/assets/89186/7550458/424c87dc-f66c-11e4-8f78-9d4f26302ef0.png)

Upgrade started at 16:20, last node was updated at 16:31. Revert to 0.8.1-RC1 happened at 16:41.

I mentioned performance in #1472 as well.

**Comment by kolloch** [comment]
Hi @bobrik,

thanks for reporting this. With which Mesos library version (native) are you using with Marathon? Note that we updated the Dockerfile to use the latest Mesos-Libraries just today because the corresponding base image was just recently released.

@lloesche have you seen something similar?

**Comment by bobrik** [comment]
Turns out that I was using 0.22.0 as a base image, will try with 0.22.1.

**Comment by bobrik** [comment]
Nah, it's still bad:

![image](https://cloud.githubusercontent.com/assets/89186/7570338/025a5fc0-f818-11e4-819f-16456482dd60.png)

Elasticsearch has [hot_threads](http://www.elastic.co/guide/en/elasticsearch/reference/1.3/cluster-nodes-hot-threads.html) api, do you have something similar so I can give more meaningful data?

**Comment by kolloch** [comment]
You can use poor man's profiling tool: Could you use jstack a couple of times on the Marathon process and send us the stack traces?

```bash
> jstack <MARATHON_PID> >stackX.txt
> jstack <MARATHON_PID> >stackX.txt
> jstack <MARATHON_PID> >stackX.txt
...
```

You might have to enter the process space with `docker exec` to do that.

**Comment by bobrik** [comment]
Here it is: https://gist.github.com/bobrik/87b8903cc3d502afe888

**Comment by drexin** [comment]
@bobrik I couldn't reproduce this. I ran v0.8.2-RC2 in a docker and started 100 tasks without the Marathon process using considerably more than 10% CPU. What does your setup look like?

**Comment by bobrik** [comment]
Marathon is running on `Intel(R) Xeon(R) CPU E3-1230 V2 @ 3.30GHz`, 8 cores.

Health checks are running once per 3-5 seconds.

![image](https://cloud.githubusercontent.com/assets/89186/7585654/9dc0b65e-f8ad-11e4-827e-4d8a5daf4bf4.png)

**Comment by drexin** [comment]
@bobrik Woud it be possible for you to change the health checks to COMMAND checks that call curl and see if the CPU usage is still that high then?

**Comment by bobrik** [comment]
I tried, but it didn't work:

```yaml
        healthChecks:
          - protocol: COMMAND
            command:
              value: "curl -f -X GET http://$HOST:$PORT0/?n=marathon_healthcheck"
            gracePeriodSeconds: 15
            maxConsecutiveFailures: 300
            intervalSeconds: 2
            timeoutSeconds: 5
```

```
{"log":"[2015-05-12 21:26:54,159] INFO Received status update for task topface_lenny-test.67ca9fbb-f8ec-11e4-ab20-56847afe9799: TASK_RUNNING () (mesosphere.marathon.MarathonScheduler:148)\n","stream":"stdout","time":"2015-05-12T21:26:54.159345729Z"}
{"log":"[2015-05-12 21:26:54,160] INFO Received status for [topface_lenny-test.67ca9fbb-f8ec-11e4-ab20-56847afe9799] with version [2015-05-12T21:18:15.625Z] and healthy [false] (mesosphere.marathon.health.MarathonHealthCheckManager:150)\n","stream":"stdout","time":"2015-05-12T21:26:54.160155174Z"}
{"log":"[2015-05-12 21:26:54,160] INFO Forwarding health result [Unhealthy(topface_lenny-test.67ca9fbb-f8ec-11e4-ab20-56847afe9799,2015-05-12T21:18:15.625Z,,2015-05-12T21:26:54.160Z)] to health check actor [Actor[akka://marathon/user/$N#1518643597]] (mesosphere.marathon.health.MarathonHealthCheckManager:171)\n","stream":"stdout","time":"2015-05-12T21:26:54.1603141Z"}
```

No more info is provided to resolve the issue. Task is healthy and works with http check. Can be related to #1380.

**Comment by bobrik** [comment]
```yaml
        healthChecks:
          - protocol: COMMAND
            command:
              value: "true"
            gracePeriodSeconds: 15
            maxConsecutiveFailures: 300
            intervalSeconds: 2
            timeoutSeconds: 5
```

120 instances, 0.8.1-RC1. Not sure if upgrading to master wouldn't kill marathon completely.

![image](https://cloud.githubusercontent.com/assets/89186/7599102/18bd73e2-f908-11e4-9555-bc70e30d810d.png)

**Comment by bobrik** [comment]
Probably worth mentioning: each of 3 marathons receives 1 rps for `/v2/apps?embed=apps.tasks` for service discovery reasons, in total it's 3 rps to master.

**Comment by kolloch** [comment]
I guess these queries are more or less the only thing I see in the stack traces that sticks out:

```
"qtp265226115-746" prio=10 tid=0x00007f0964023000 nid=0xd88 waiting on condition [0x00007f091bbf8000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000007dafdd118> (a scala.concurrent.impl.Promise$CompletionLatch)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:226)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedNanos(AbstractQueuedSynchronizer.java:1033)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1326)
	at scala.concurrent.impl.Promise$DefaultPromise.tryAwait(Promise.scala:208)
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:218)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)
	at scala.concurrent.Await$$anonfun$result$1.apply(package.scala:190)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
	at scala.concurrent.Await$.result(package.scala:190)
	at mesosphere.marathon.api.RestResource$class.result(RestResource.scala:44)
	at mesosphere.marathon.api.v2.AppsResource.result(AppsResource.scala:29)
	at mesosphere.marathon.api.v2.AppsResource.mesosphere$marathon$api$v2$AppsResource$$enrichedTasks(AppsResource.scala:294)
	at mesosphere.marathon.api.v2.AppsResource$$anonfun$5.apply(AppsResource.scala:57)
	at mesosphere.marathon.api.v2.AppsResource$$anonfun$5.apply(AppsResource.scala:55)
	at scala.collection.immutable.Stream$$anonfun$map$1.apply(Stream.scala:418)
	at scala.collection.immutable.Stream$$anonfun$map$1.apply(Stream.scala:418)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1222)
	- locked <0x00000007dafdbf40> (a scala.collection.immutable.Stream$Cons)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1212)
	at scala.collection.immutable.Stream$$anonfun$map$1.apply(Stream.scala:418)
	at scala.collection.immutable.Stream$$anonfun$map$1.apply(Stream.scala:418)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1222)
	- locked <0x00000007dafdbf90> (a scala.collection.immutable.Stream$Cons)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1212)
	at scala.collection.immutable.Stream.foreach(Stream.scala:595)
	at play.api.libs.json.JsValueSerializer.serialize(JsValue.scala:311)
	at play.api.libs.json.JsValueSerializer$$anonfun$serialize$2.apply(JsValue.scala:320)
	at play.api.libs.json.JsValueSerializer$$anonfun$serialize$2.apply(JsValue.scala:318)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at play.api.libs.json.JsValueSerializer.serialize(JsValue.scala:318)
	at play.api.libs.json.JsValueSerializer.serialize(JsValue.scala:302)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:114)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:1887)
	at play.api.libs.json.JacksonJson$.generateFromJsValue(JsValue.scala:495)
	at play.api.libs.json.Json$.stringify(Json.scala:51)
	at play.api.libs.json.JsValue$class.toString(JsValue.scala:80)
	at play.api.libs.json.JsObject.toString(JsValue.scala:166)
	at mesosphere.marathon.api.v2.AppsResource.index(AppsResource.scala:87)
	at sun.reflect.GeneratedMethodAccessor54.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.sun.jersey.spi.container.JavaMethodInvokerFactory$1.invoke(JavaMethodInvokerFactory.java:60)
	at com.sun.jersey.server.impl.model.method.dispatch.AbstractResourceMethodDispatchProvider$TypeOutInvoker._dispatch(AbstractResourceMethodDispatchProvider.java:185)
	at com.sun.jersey.server.impl.model.method.dispatch.ResourceJavaMethodDispatcher.dispatch(ResourceJavaMethodDispatcher.java:75)
	at com.codahale.metrics.jersey.InstrumentedResourceMethodDispatchProvider$TimedRequestDispatcher.dispatch(InstrumentedResourceMethodDispatchProvider.java:30)
	at com.sun.jersey.server.impl.uri.rules.HttpMethodRule.accept(HttpMethodRule.java:302)
	at com.sun.jersey.server.impl.uri.rules.ResourceClassRule.accept(ResourceClassRule.java:108)
	at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147)
	at com.sun.jersey.server.impl.uri.rules.RootResourceClassesRule.accept(RootResourceClassesRule.java:84)
	at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1542)
	at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1473)
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419)
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409)
	at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409)
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:540)
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:715)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:848)
	at com.google.inject.servlet.ServletDefinition.doService(ServletDefinition.java:263)
	at com.google.inject.servlet.ServletDefinition.service(ServletDefinition.java:178)
	at com.google.inject.servlet.ManagedServletPipeline.service(ManagedServletPipeline.java:91)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:62)
	at mesosphere.marathon.api.CacheDisablingFilter.doFilter(CacheDisablingFilter.scala:18)
	at com.google.inject.servlet.FilterDefinition.doFilter(FilterDefinition.java:163)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:58)
	at mesosphere.marathon.api.CORSFilter.doFilter(CORSFilter.scala:46)
	at com.google.inject.servlet.FilterDefinition.doFilter(FilterDefinition.java:163)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:58)
	at mesosphere.marathon.api.LeaderProxyFilter.doFilter(LeaderProxyFilter.scala:56)
	at com.google.inject.servlet.FilterDefinition.doFilter(FilterDefinition.java:163)
	at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:58)
	at com.google.inject.servlet.ManagedFilterPipeline.dispatch(ManagedFilterPipeline.java:118)
	at com.google.inject.servlet.GuiceFilter.doFilter(GuiceFilter.java:113)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1467)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:501)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1086)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:429)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1020)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:135)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)
	at com.codahale.metrics.jetty8.InstrumentedHandler.handle(InstrumentedHandler.java:192)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:154)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)
	at org.eclipse.jetty.server.Server.handle(Server.java:370)
	at org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:494)
	at org.eclipse.jetty.server.AbstractHttpConnection.headerComplete(AbstractHttpConnection.java:971)
	at org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.headerComplete(AbstractHttpConnection.java:1033)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:644)
	at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:235)
	at org.eclipse.jetty.server.AsyncHttpConnection.handle(AsyncHttpConnection.java:82)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:696)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:608)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:543)
	at java.lang.Thread.run(Thread.java:745)
```

I don't know yet why this would have changed recently.

**Comment by bobrik** [comment]
I've shut down service discovery updater for a few minutes on 0.8.1-RC1 and look at that:

![image](https://cloud.githubusercontent.com/assets/89186/7612056/fdbc125e-f990-11e4-8dea-8f03e8b00be7.png)

The second decrease in cpu usage on the graph happened when I closed browser tab with marathon.

**Comment by bobrik** [comment]
Flamegraphs: https://gist.github.com/bobrik/969d322bb28c6a649cf7

https://github.com/jrudolph/perf-map-agent


0.8.1-RC1:

![flamegraph-0.8.1-rc1](https://cloud.githubusercontent.com/assets/89186/7629582/724d4c8c-fa36-11e4-9795-fea21900a0de.png)

0.8.2-SNAPSHOT:

![flamegraph-0.8.2-snapshot](https://cloud.githubusercontent.com/assets/89186/7629585/76d281a0-fa36-11e4-845a-541f7e57583f.png)

Blue line, higher is 0.8.2:

![image](https://cloud.githubusercontent.com/assets/89186/7629634/f19d7ab6-fa36-11e4-9110-0d2b88c7f63c.png)

**Comment by dgromov** [comment]
@drexin, were you using OpenJDK when you tried reproducing this? I wonder if that had something to do with it. https://gist.github.com/bobrik/87b8903cc3d502afe888 suggests that these numbers are all using that.

**Comment by apuckey** [comment]
I'm using Sun Java 1.7 not openjdk if that makes a difference

**Comment by kolloch** [comment]
Hi @bobrik, BTW I hope that it is clear that we really appreciate your detailed reporting. Unfortunately, we can't reproduce it so far.

I could imagine that it has something to do with the Mesos Library changes. Did you try 0.8.1-RC1 with the 0.22.1 mesos libraries by any chance? I am not sure if that is a supported configuration but if that exhibits the same CPU pattern, the reason could lie in the new Mesos Library version.

I think it might make sense to implement #1539 soon and check if your problems persist.

What do you think?

**Comment by bobrik** [comment]
Should I just try 0.8.1-RC1 tag on top of `mesosphere/mesos:0.22.1` docker image?

Removing native code sounds like a good idea, too much is happening there.

**Comment by kolloch** [comment]
Hi @bobrik, if it's not a big hassle (at least in comparison to the things you have already done), trying 0.8.1-RC1 on top of mesosphere/mesos:0.22.1 would be grand. :+1:

**Comment by bobrik** [comment]
Looks like marathon itself is to blame and 0.22.1 is even better with 0.8.2-SNAPSHOT. Snapshot is from today's master, btw.

~ 12:10 marathon 0.8.1-RC1, mesos 0.22.1
~ 12:25 marathon 0.8.1-RC1, mesos 0.22.0
~ 12:38 marathon 0.8.2-SNAPSHOT, mesos 0.22.1
~ 12:47 marathon 0.8.2-SNAPSHOT, mesos 0.22.0

![image](https://cloud.githubusercontent.com/assets/89186/7700454/7659c108-fe27-11e4-9a96-922ebb876db9.png)

**Comment by kolloch** [comment]
Ok, I will try to reproduce that by creating

* ~50 apps with ~5 tasks each.
* querying `/v2/apps?embed=apps.tasks` 3 times per second

I will test this with the last release candidate and the current release candidate and check CPU usage. I will use the docker images to have a hopefully reproducible test base.

Any other suggestions?

**Comment by bobrik** [comment]
You can go with just 1 app, the number of instance matters. Here I scaled app from 5 to 100 instances:

![image](https://cloud.githubusercontent.com/assets/89186/7701253/00c33102-fe2e-11e4-8d9b-90b15c1fbe92.png)

Don't forget about frequent health checks as well.

**Comment by kolloch** [comment]
Hi @bobrik, I tried to reproduce the problem and I failed so far. Actually, as far as I can see the performance has improved (by a lot).

I tried to describe my hopefully reproducible test setup here:

https://gist.github.com/kolloch/e988be8fd631b9696729

So I guess there is still something important that's different in your environment. Did I miss something crucial?

**Comment by bobrik** [comment]
Interesting, I've tried test cluster in docker compose with 1 master, 1 slave, 1 marathon and 50 containers and it looks like 0.8.3-RC3 is indeed better: 20rps vs 15rps. I'll try to figure out what's missing.

**Comment by bobrik** [comment]
I was trying to set separate Marathon cluster in production Mesos cluster, but now I'm getting segfaults all the time. Marathon cannot even register as a framework:

```
# docker run --rm -it --net host mesosphere/marathon:v0.8.2-RC3 --master zk://web488:2181,web489:2181,web490:2181/mesos --zk zk://web488:2181,web489:2181,web490:2181/marathonugh --hostname web13
MESOS_NATIVE_JAVA_LIBRARY is not set. Searching in /usr/lib /usr/local/lib.
MESOS_NATIVE_LIBRARY, MESOS_NATIVE_JAVA_LIBRARY set to '/usr/lib/libmesos.so'
[2015-05-20 13:36:25,052] INFO Starting Marathon 0.8.2-RC3 (mesosphere.marathon.Main$:87)
[2015-05-20 13:36:26,778] INFO Connecting to Zookeeper... (mesosphere.marathon.Main$:37)
[2015-05-20 13:36:26,789] INFO Client environment:zookeeper.version=3.3.3-1203054, built on 11/17/2011 05:47 GMT (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,789] INFO Client environment:host.name=web13 (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,789] INFO Client environment:java.version=1.7.0_79 (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,789] INFO Client environment:java.vendor=Oracle Corporation (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,789] INFO Client environment:java.home=/usr/lib/jvm/java-7-openjdk-amd64/jre (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,789] INFO Client environment:java.class.path=./bin/../target/marathon-assembly-0.8.2-RC3.jar (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,789] INFO Client environment:java.library.path=/usr/java/packages/lib/amd64:/usr/lib/x86_64-linux-gnu/jni:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/usr/lib/jni:/lib:/usr/lib (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,789] INFO Client environment:java.io.tmpdir=/tmp (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,790] INFO Client environment:java.compiler=<NA> (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,790] INFO Client environment:os.name=Linux (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,790] INFO Client environment:os.arch=amd64 (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,790] INFO Client environment:os.version=3.14.18-gentoo (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,790] INFO Client environment:user.name=root (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,790] INFO Client environment:user.home=/root (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,792] INFO Client environment:user.dir=/marathon (org.apache.zookeeper.ZooKeeper:97)
[2015-05-20 13:36:26,793] INFO Initiating client connection, connectString=web488:2181,web489:2181,web490:2181 sessionTimeout=10000 watcher=com.twitter.common.zookeeper.ZooKeeperClient$3@51b9d4c (org.apache.zookeeper.ZooKeeper:379)
[2015-05-20 13:36:26,803] INFO Opening socket connection to server web490/192.168.1.234:2181 (org.apache.zookeeper.ClientCnxn:1061)
[2015-05-20 13:36:26,809] INFO Socket connection established to web490/192.168.1.234:2181, initiating session (org.apache.zookeeper.ClientCnxn:950)
[2015-05-20 13:36:26,814] INFO Session establishment complete on server web490/192.168.1.234:2181, sessionid = 0x34d6c198608003c, negotiated timeout = 10000 (org.apache.zookeeper.ClientCnxn:739)
2015-05-20 13:36:27,838:1(0x7ff66641c700):ZOO_INFO@log_env@712: Client environment:zookeeper.version=zookeeper C client 3.4.5
2015-05-20 13:36:27,838:1(0x7ff66641c700):ZOO_INFO@log_env@716: Client environment:host.name=web13
2015-05-20 13:36:27,838:1(0x7ff66641c700):ZOO_INFO@log_env@723: Client environment:os.name=Linux
2015-05-20 13:36:27,838:1(0x7ff66641c700):ZOO_INFO@log_env@724: Client environment:os.arch=3.14.18-gentoo
2015-05-20 13:36:27,838:1(0x7ff66641c700):ZOO_INFO@log_env@725: Client environment:os.version=#2 SMP Fri Sep 19 17:58:42 MSK 2014
2015-05-20 13:36:27,838:1(0x7ff66641c700):ZOO_INFO@log_env@733: Client environment:user.name=root
2015-05-20 13:36:27,838:1(0x7ff66641c700):ZOO_INFO@log_env@741: Client environment:user.home=/root
2015-05-20 13:36:27,838:1(0x7ff66641c700):ZOO_INFO@log_env@753: Client environment:user.dir=/marathon
2015-05-20 13:36:27,838:1(0x7ff66641c700):ZOO_INFO@zookeeper_init@786: Initiating client connection, host=web488:2181,web489:2181,web490:2181 sessionTimeout=10000 watcher=0x7ff666e35a60 sessionId=0 sessionPasswd=<null> context=0x7ff5f8000b40 flags=0
2015-05-20 13:36:27,840:1(0x7ff65dffb700):ZOO_INFO@check_events@1703: initiated connection to server [192.168.1.234:2181]
2015-05-20 13:36:27,842:1(0x7ff65dffb700):ZOO_INFO@check_events@1750: session establishment complete on server [192.168.1.234:2181], sessionId=0x34d6c198608003d, negotiated timeout=10000
[2015-05-20 13:36:28,120] INFO Registering in Zookeeper with hostname:web13 (mesosphere.marathon.MarathonModule:140)
[2015-05-20 13:36:28,200] INFO Adding HTTP support. (mesosphere.marathon.MarathonApp$$anon$1:47)
[2015-05-20 13:36:28,201] INFO No HTTPS support configured. (mesosphere.marathon.MarathonApp$$anon$1:50)
[2015-05-20 13:36:28,206] INFO Starting up (mesosphere.marathon.MarathonSchedulerService:148)
[2015-05-20 13:36:28,206] INFO Beginning run (mesosphere.marathon.MarathonSchedulerService:153)
[2015-05-20 13:36:28,207] INFO Will offer leadership after 500 milliseconds backoff (mesosphere.marathon.MarathonSchedulerService:334)
[2015-05-20 13:36:28,208] INFO jetty-8.y.z-SNAPSHOT (org.eclipse.jetty.server.Server:272)
[2015-05-20 13:36:28,343] INFO Registering com.codahale.metrics.jersey.InstrumentedResourceMethodDispatchAdapter as a provider class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:113)
[2015-05-20 13:36:28,345] INFO Registering com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider as a provider class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:113)
[2015-05-20 13:36:28,346] INFO Registering mesosphere.chaos.validation.ConstraintViolationExceptionMapper as a provider class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:113)
[2015-05-20 13:36:28,346] INFO Registering mesosphere.marathon.api.MarathonExceptionMapper as a provider class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:113)
[2015-05-20 13:36:28,346] INFO Registering mesosphere.marathon.api.v2.AppsResource as a root resource class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:116)
[2015-05-20 13:36:28,346] INFO Registering mesosphere.marathon.api.v2.TasksResource as a root resource class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:116)
[2015-05-20 13:36:28,346] INFO Registering mesosphere.marathon.api.v2.EventSubscriptionsResource as a root resource class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:116)
[2015-05-20 13:36:28,347] INFO Registering mesosphere.marathon.api.v2.QueueResource as a root resource class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:116)
[2015-05-20 13:36:28,347] INFO Registering mesosphere.marathon.api.v2.GroupsResource as a root resource class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:116)
[2015-05-20 13:36:28,347] INFO Registering mesosphere.marathon.api.v2.InfoResource as a root resource class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:116)
[2015-05-20 13:36:28,347] INFO Registering mesosphere.marathon.api.v2.LeaderResource as a root resource class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:116)
[2015-05-20 13:36:28,347] INFO Registering mesosphere.marathon.api.v2.DeploymentsResource as a root resource class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:116)
[2015-05-20 13:36:28,347] INFO Registering mesosphere.marathon.api.v2.ArtifactsResource as a root resource class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:116)
[2015-05-20 13:36:28,348] INFO Registering mesosphere.marathon.api.v2.SchemaResource as a root resource class (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:116)
[2015-05-20 13:36:28,350] INFO Initiating Jersey application, version 'Jersey: 1.18.1 02/19/2014 03:28 AM' (com.sun.jersey.server.impl.application.WebApplicationImpl:815)
[2015-05-20 13:36:28,411] INFO Binding com.codahale.metrics.jersey.InstrumentedResourceMethodDispatchAdapter to GuiceManagedComponentProvider with the scope "Singleton" (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:168)
[2015-05-20 13:36:28,426] INFO Binding mesosphere.chaos.validation.ConstraintViolationExceptionMapper to GuiceManagedComponentProvider with the scope "Singleton" (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:168)
[2015-05-20 13:36:28,426] INFO Binding mesosphere.marathon.api.MarathonExceptionMapper to GuiceManagedComponentProvider with the scope "Singleton" (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:168)
[2015-05-20 13:36:28,427] INFO Binding com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider to GuiceManagedComponentProvider with the scope "Singleton" (com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory:168)
[2015-05-20 13:36:28,445] INFO HV000001: Hibernate Validator 5.1.2.Final (org.hibernate.validator.internal.util.Version:27)
[2015-05-20 13:36:28,727] INFO Using HA and therefore offering leadership (mesosphere.marathon.MarathonSchedulerService:341)
[2015-05-20 13:36:28,763] INFO Set group member ID to member_0000000000 (com.twitter.common.zookeeper.Group:426)
[2015-05-20 13:36:28,776] INFO Candidate /marathonugh/leader/member_0000000000 is now leader of group: [member_0000000000] (com.twitter.common.zookeeper.CandidateImpl:152)
[2015-05-20 13:36:28,776] INFO Elected (Leader Interface) (mesosphere.marathon.MarathonSchedulerService:252)
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007ff688b8fe0c, pid=1, tid=140696219645696
#
# JRE version: OpenJDK Runtime Environment (7.0_79-b14) (build 1.7.0_79-b14)
# Java VM: OpenJDK 64-Bit Server VM (24.79-b02 mixed mode linux-amd64 compressed oops)
# Derivative: IcedTea 2.5.5
# Distribution: Ubuntu 14.04 LTS, package 7u79-2.5.5-0ubuntu0.14.04.2
# Problematic frame:
# C  [libc.so.6+0x82e0c]  cfree+0x1c
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /marathon/hs_err_pid1.log
#
# If you would like to submit a bug report, please include
# instructions on how to reproduce the bug and visit:
#   http://icedtea.classpath.org/bugzilla
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

[error occurred during error reporting , id 0xb]


[error occurred during error reporting , id 0xb]


[error occurred during error reporting , id 0xb]


[error occurred during error reporting , id 0xb]
```

**Comment by kolloch** [comment]
Hi @bobrik, can you open a separate issue for that please?

**Comment by bobrik** [comment]
I can't reproduce it outside of main Marathon cluster. It'd be super helpful if mesosphere provided debug symbols for `mesos` package so I can build image with debug tools for jvm and make proper flamegraph.

I found 1137 versions of root group in marathon's znode (#1533 probably), but leaving just one and restart didn't help. Setting version history limit to 10 and redeploying apps that had more version didn't help either.

I don't want to move apps to the new cluster, it might end up just like #1082 and I have much more production apps now. I can give you app definitions for deployed apps, though, if that could help.

SIGSEGV only happened on two nodes, I just picked fresh servers and it worked. Not sure how you can reproduce that. Debug symbols for `mesos` package could give a meaningful core dump, though. #1352 can be related.

Let me know if you have a good idea how to get to the bottom of this issue.

**Comment by kolloch** [comment]
@joerg84 Could you help with using the appropriate debug symbols, please? Can we enable better logging in the native code? Thanks.

For the deteriorated CPU consumption maybe Java profiling can help if the code hotspots are in the Java code. If you can access the machine and have a Oracle JDK installed, you could use [Java Mission Control](http://www.oracle.com/technetwork/java/javaseproducts/mission-control/java-mission-control-1998576.html) to look at the hotspots (the jmc command). That will, of course, decrease performance of the process under observation further, even though sampling isn't to bad.  Or use the simple [hprof](http://docs.oracle.com/javase/7/docs/technotes/samples/hprof.html) mechanism .

**Comment by bobrik** [comment]
```
CPU SAMPLES BEGIN (total = 266814) Thu May 21 09:51:33 2015
rank   self  accum   count trace method
   1 57.67% 57.67%  153882 300643 sun.nio.ch.EPollArrayWrapper.epollWait
   2 16.79% 74.46%   44788 300911 sun.nio.ch.ServerSocketChannelImpl.accept0
   3  8.56% 83.02%   22830 301094 org.apache.mesos.state.AbstractState$FetchFuture.get
   4  7.37% 90.38%   19651 301099 org.apache.mesos.MesosSchedulerDriver.join
   5  4.02% 94.41%   10738 301100 org.apache.mesos.state.AbstractState$NamesFuture.get
   6  1.80% 96.20%    4792 300955 java.net.SocketInputStream.socketRead0
   7  0.37% 96.58%     997 300915 java.util.zip.ZipFile.getNextEntry
   8  0.19% 96.77%     508 301105 mesosphere.marathon.state.AppDefinition.mergeFromProto
   9  0.14% 96.91%     372 300968 java.util.zip.ZipFile.getEntryBytes
  10  0.14% 97.04%     363 300917 java.util.zip.ZipFile.getEntryTime
  11  0.10% 97.15%     280 300967 java.util.zip.ZipFile.freeEntry
  12  0.10% 97.25%     280 301263 spray.httpx.RequestBuilding$RequestBuilder.apply
  13  0.10% 97.35%     262 301227 sun.misc.Unsafe.unpark
  14  0.10% 97.45%     255 300919 java.util.zip.ZipFile$1.hasMoreElements
  15  0.08% 97.53%     215 300171 java.util.zip.ZipFile.read
  16  0.08% 97.60%     207 301275 spray.httpx.RequestBuilding$RequestBuilder.apply
  17  0.07% 97.68%     191 301523 sun.misc.Unsafe.park
  18  0.06% 97.74%     172 301209 java.util.regex.Pattern$BnM.optimize
  19  0.05% 97.79%     129 300181 java.lang.ClassLoader.defineClass1
  20  0.04% 97.83%     112 301192 sun.misc.Unsafe.park
```

http://www.brendangregg.com/blog/2014-06-09/java-cpu-sampling-using-hprof.html

**Comment by kolloch** [comment]
Thanks a lot @bobrik. Since the first two entries actually signify idle threads, it really seems to be related to the state/zookeeper interaction.

Just an idea: Do you Marathon clients maybe just perform requests as fast as possible and thus query Marathon more often if it answers faster?

Can you put the output of getting `/metrics` to some gist for analysis? Maybe for a Marathon with the `--enable_metrics` flag (enables additional metrics)? It would be nice to have them for the old version as well.

Maybe you could also somehow grep the HTTP requests to Marathon (GET / ...)?

**Comment by bobrik** [comment]
Requests to marathon are synchronous, I don't send the next one before previous returns. There is also a pause of 1 second.

I tried to get metrics under load and found that 0.8.2-RC3 today shows only slightly higher load under regular conditions (~580 ms vs ~500ms):

![image](https://cloud.githubusercontent.com/assets/89186/7747478/7879a096-ffc3-11e4-800c-d6c5c2bacddc.png)

Maybe marathon somehow took advantage of removed versions from zk, but it was yesterday.

Sorry for 1-minute granularity of my monitoring, I'm still fighting with people to get 10-second resolution.

1. Before 13:35 there was 0.8.2-RC3
2. Then 0.8.1-RC1 with 2 load tests at 13:44 (results below)
3. Then 0.8.2-RC3 with 2 load tests at 13:54 (results below)
4. Then 0.8.3-RC3 with hprof on master node at 14:02 (results below)

Load tests for 2 (0.8.1-RC1):

```
Server Software:        Jetty(8.y.z-SNAPSHOT)
Server Hostname:        web489
Server Port:            8080

Document Path:          /v2/apps?embed=apps.tasks
Document Length:        113643 bytes

Concurrency Level:      10
Time taken for tests:   59.010 seconds
Complete requests:      1000
Failed requests:        0
Write errors:           0
Total transferred:      113807000 bytes
HTML transferred:       113643000 bytes
Requests per second:    16.95 [#/sec] (mean)
Time per request:       590.097 [ms] (mean)
Time per request:       59.010 [ms] (mean, across all concurrent requests)
Transfer rate:          1883.41 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       1
Processing:   221  589 190.7    558    1741
Waiting:      220  584 189.9    553    1736
Total:        222  589 190.7    558    1741

Percentage of the requests served within a certain time (ms)
  50%    558
  66%    634
  75%    687
  80%    729
  90%    824
  95%    930
  98%   1091
  99%   1214
 100%   1741 (longest request)
```

Load tests for 3 (0.8.2-RC3):

```
Server Software:        Jetty(8.y.z-SNAPSHOT)
Server Hostname:        web489
Server Port:            8080

Document Path:          /v2/apps?embed=apps.tasks
Document Length:        115178 bytes

Concurrency Level:      10
Time taken for tests:   166.017 seconds
Complete requests:      1000
Failed requests:        0
Write errors:           0
Total transferred:      115342000 bytes
HTML transferred:       115178000 bytes
Requests per second:    6.02 [#/sec] (mean)
Time per request:       1660.173 [ms] (mean)
Time per request:       166.017 [ms] (mean, across all concurrent requests)
Transfer rate:          678.48 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       1
Processing:   445 1656 340.6   1637    3095
Waiting:      443 1655 340.5   1636    3091
Total:        445 1656 340.6   1637    3095

Percentage of the requests served within a certain time (ms)
  50%   1637
  66%   1755
  75%   1843
  80%   1903
  90%   2060
  95%   2243
  98%   2540
  99%   2732
 100%   3095 (longest request)
```

Load tests for 4 (0.8.2-RC3 with hprof):

```
Server Software:        Jetty(8.y.z-SNAPSHOT)
Server Hostname:        web489
Server Port:            8080

Document Path:          /v2/apps?embed=apps.tasks
Document Length:        115178 bytes

Concurrency Level:      10
Time taken for tests:   163.036 seconds
Complete requests:      1000
Failed requests:        0
Write errors:           0
Total transferred:      115342000 bytes
HTML transferred:       115178000 bytes
Requests per second:    6.13 [#/sec] (mean)
Time per request:       1630.362 [ms] (mean)
Time per request:       163.036 [ms] (mean, across all concurrent requests)
Transfer rate:          690.88 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       1
Processing:   438 1628 251.4   1607    2566
Waiting:      433 1626 251.3   1605    2564
Total:        438 1628 251.4   1607    2566

Percentage of the requests served within a certain time (ms)
  50%   1607
  66%   1691
  75%   1770
  80%   1807
  90%   1926
  95%   2077
  98%   2246
  99%   2357
 100%   2566 (longest request)
```

Top traces from hprof:

```
TRACE 301009:
	org.apache.mesos.state.AbstractState$NamesFuture.get(AbstractState.java:Unknown line)
	org.apache.mesos.state.AbstractState$NamesFuture.get(AbstractState.java:355)
	mesosphere.util.BackToTheFuture$$anonfun$futureToFuture$1$$anonfun$apply$2.apply(BackToTheFuture.scala:35)
	scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
TRACE 300997:
	org.apache.mesos.state.AbstractState$FetchFuture.get(AbstractState.java:Unknown line)
	org.apache.mesos.state.AbstractState$FetchFuture.get(AbstractState.java:226)
	mesosphere.util.BackToTheFuture$$anonfun$futureToFutureOption$1$$anonfun$apply$1.apply(BackToTheFuture.scala:22)
	mesosphere.util.BackToTheFuture$$anonfun$futureToFutureOption$1$$anonfun$apply$1.apply(BackToTheFuture.scala:21)
TRACE 300627:
	sun.nio.ch.EPollArrayWrapper.epollWait(EPollArrayWrapper.java:Unknown line)
	sun.nio.ch.EPollArrayWrapper.poll(EPollArrayWrapper.java:269)
	sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:79)
	sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:87)
TRACE 300900:
	sun.nio.ch.ServerSocketChannelImpl.accept0(ServerSocketChannelImpl.java:Unknown line)
	sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:241)
	org.eclipse.jetty.server.nio.SelectChannelConnector.accept(SelectChannelConnector.java:109)
	org.eclipse.jetty.server.AbstractConnector$Acceptor.run(AbstractConnector.java:938)
TRACE 301004:
	org.apache.mesos.MesosSchedulerDriver.join(MesosSchedulerDriver.java:Unknown line)
	org.apache.mesos.MesosSchedulerDriver.run(MesosSchedulerDriver.java:255)
	mesosphere.marathon.MarathonSchedulerService$$anonfun$runDriver$1$$anonfun$apply$mcV$sp$1.apply(MarathonSchedulerService.scala:192)
	mesosphere.marathon.MarathonSchedulerService$$anonfun$runDriver$1$$anonfun$apply$mcV$sp$1.apply(MarathonSchedulerService.scala:192)
TRACE 300904:
	java.util.zip.ZipFile.getNextEntry(ZipFile.java:Unknown line)
	java.util.zip.ZipFile.access$500(ZipFile.java:56)
	java.util.zip.ZipFile$1.nextElement(ZipFile.java:497)
	java.util.zip.ZipFile$1.nextElement(ZipFile.java:483)
TRACE 300943:
	java.net.SocketInputStream.socketRead0(SocketInputStream.java:Unknown line)
	java.net.SocketInputStream.read(SocketInputStream.java:152)
	java.net.SocketInputStream.read(SocketInputStream.java:122)
	java.io.BufferedInputStream.fill(BufferedInputStream.java:235)
```

```
CPU SAMPLES BEGIN (total = 901878) Thu May 21 11:10:25 2015
rank   self  accum   count trace method
   1 35.96% 35.96%  324324 301009 org.apache.mesos.state.AbstractState$NamesFuture.get
   2 30.68% 66.64%  276697 300997 org.apache.mesos.state.AbstractState$FetchFuture.get
   3 19.98% 86.62%  180191 300627 sun.nio.ch.EPollArrayWrapper.epollWait
   4  5.09% 91.71%   45901 300900 sun.nio.ch.ServerSocketChannelImpl.accept0
   5  4.73% 96.44%   42631 301004 org.apache.mesos.MesosSchedulerDriver.join
   6  0.45% 96.88%    4030 300904 java.util.zip.ZipFile.getNextEntry
   7  0.22% 97.11%    2011 300943 java.net.SocketInputStream.socketRead0
   8  0.17% 97.27%    1516 300955 java.util.zip.ZipFile.getEntryTime
   9  0.15% 97.43%    1392 300956 java.util.zip.ZipFile.getEntryBytes
  10  0.15% 97.57%    1309 301014 sun.misc.Unsafe.unpark
  11  0.12% 97.70%    1099 300954 java.util.zip.ZipFile.freeEntry
  12  0.10% 97.80%     892 301189 spray.httpx.RequestBuilding$RequestBuilder.apply
  13  0.07% 97.87%     642 301035 org.apache.mesos.state.AbstractState$FetchFuture.get
  14  0.07% 97.94%     627 300907 java.util.zip.ZipFile$1.hasMoreElements
  15  0.07% 98.01%     627 301011 java.util.regex.Pattern$BnM.optimize
  16  0.06% 98.07%     566 301075 sun.misc.Unsafe.park
  17  0.06% 98.13%     523 301316 org.apache.mesos.state.AbstractState.__names
  18  0.05% 98.17%     434 300049 java.util.zip.ZipFile.read
  19  0.04% 98.22%     397 301013 mesosphere.marathon.state.AppDefinition.mergeFromProto
```

I don't really understand why it takes so much effort to serialize internal in-memory state to json and return it back to the caller. What else is happening there?

Do you need metrics after some load testing or metrics for regular load are fine too?

**Comment by kolloch** [comment]
Actually the deserialization happens in the Scala (JVM) code (from bytes to objects) so we should see that in the profiling if that mattered. The code seems to mostly wait for the responses from Zookeeper (accesses via the mesos library). Either the library has got slower (something which we kind of ruled out because you said using 0.22.1 actually sped up the whole thing slightly) or we are performing more requests.

The metrics would show invocation counts. It would be nice if the metrics for the old/new version were somewhat comparable. So collected under similar conditions for a similar time interval. If the problem already shows under normal load, we do not need any load even though it might make things clearer.

**Comment by bobrik** [comment]
Ok, I'll try to collect metrics from master on 0.8.1-RC1 and 0.8.2-RC3 with `--enable_metrics` after 30 minutes of regular load.

Meanwhile, can you tell me what is needed from zk when I ask for `/v2/apps?embed=apps.tasks`? I thought that marathon keeps everything in memory even though state is written to zk for recovery.

**Comment by kolloch** [comment]
Hi @bobrik, actually, reads currently go to Zookeeper as well. We want to change that. Basically, it is also a trade-off between looking at current user problems (which needs time) and rewriting some of the code (which needs time) which might actually solve these issues anyway.

So, without wanting to sound smart, analyzing this issue actually prevents me from rewriting code. But I do not like to release 0.8.2 before we understand the implications.

**Comment by bobrik** [comment]
Here are the graphs from zk cluster, looks suspicious if you ask me:

![image](https://cloud.githubusercontent.com/assets/89186/7748003/3ab7aa86-ffc9-11e4-8b82-94433e72e81f.png)

There is a lot smaller load from 0.8.1-RC1, especially in terms of bytes per second.

**Comment by kolloch** [comment]
Hi @bobrik, 

I cannot really make sense of your graphs. What's the old, what's the new version? What makes you suspicious?

Can you actually tell us the configuration parameters you start marathon with? I assume, they are the same between the old and the new version? Thanks.

**Comment by bobrik** [comment]
Sorry for not making it clear. Environment vars for marathon (in an ansible playbook):

```yaml
          MARATHON_MASTER: zk://web488:2181,web489:2181,web490:2181/mesos
          MARATHON_ZK: zk://web488:2181,web489:2181,web490:2181/marathon-new
          MARATHON_ZK_MAX_VERSIONS: 10
          MARATHON_HOSTNAME: "{{ inventory_hostname }}"
```

They are the same for both versions. 

Now to the graphs, new ones this time, hope they are more clear. Here I ran 0.8.2-RC3 for 40 minutes, then 0.8.1-RC1 for 40 minutes, then 0.8.1-RC1 on top of 0.22.1 libs for 10 minutes.

Marathon cluster:

![image](https://cloud.githubusercontent.com/assets/89186/7749369/adcd64dc-ffd5-11e4-88af-ec13cc4a30be.png)

Zookeeper cluster for marathon and mesos, same time:

![image](https://cloud.githubusercontent.com/assets/89186/7749396/d0e4f8c2-ffd5-11e4-8adb-e7b866539951.png)

Enormous difference in the used bandwidth to zookeeper is suspicious: 200 kb/s vs 5000 kb/s. CPU load and packet rate are alos higher with 0.8.2-RC3.

Metrics for this, for 0.8.2 with `--enable-metrics`: https://gist.github.com/bobrik/96997d0030338fa1dc15

Does it make sense now? Thank you for your patience.

**Comment by kolloch** [comment]
Hi @bobrik,

thanks to your extensive reporting, we found the offender. The metrics in the gist helped us out.

If you are really adventurous, you can checkout the `pk/1497_health_statuses_more_efficient` branch. Otherwise, you can wait for us to merge to `master` and release another RC.

**Comment by bobrik** [comment]
Thanks, I'll test it tomorrow. Is there an issue to remove unnecessary zk read requests?

**Comment by bobrik** [comment]
Marathon master built from `origin/pr/1568`, running from 12:09:

![image](https://cloud.githubusercontent.com/assets/89186/7832866/e4b27a48-046a-11e5-8eee-6bc2c7cc990b.png)

ZK traffic didn't change compared to 0.8.1-RC1, though. Workload is slightly different since I query more groups now.

**Comment by kolloch** [comment]
Hi @bobrik, that is surprising. Can you export the metrics for us again?

**Comment by bobrik** [comment]
Metrics after 10 minutes: https://gist.github.com/bobrik/bb8b852eb1156624a3b8

**Comment by kolloch** [comment]
I'll try to summarize the findings (correct me if I am wrong).

When 

* using the proposed fixed version instead of 0.8.1-RC1
* with the same load (which is different from the load you tested with before)

you see that

* The CPU usage goes up from ~600ms to ~740ms .
* The ZK traffic doesn't change significantly.

So the new version still uses more CPU than 0.8.1-RC1 but is otherwise fine.

The increased CPU could be potentially explained by more requests. At least in the last comparison with full metrics for some reason we saw significantly more requests against the new version, maybe because of faster response times. The new version has a mean response time of 35ms for AppResource.index compared to 126ms for the old version.

If that is correct, we would like to release a new RC with the fix.

**Comment by bobrik** [comment]
> The increased CPU could be potentially explained by more requests.

0.8.1-RC1:

```
Document Path:          /v2/apps?embed=apps.tasks
Document Length:        73417 bytes

Concurrency Level:      10
Time taken for tests:   51.741 seconds
Complete requests:      1000
Failed requests:        0
Write errors:           0
Total transferred:      73581000 bytes
HTML transferred:       73417000 bytes
Requests per second:    19.33 [#/sec] (mean)
Time per request:       517.413 [ms] (mean)
Time per request:       51.741 [ms] (mean, across all concurrent requests)
Transfer rate:          1388.76 [Kbytes/sec] received
```

PR 1568:

```
Document Path:          /v2/apps?embed=apps.tasks
Document Length:        76127 bytes

Concurrency Level:      10
Time taken for tests:   84.561 seconds
Complete requests:      1000
Failed requests:        0
Write errors:           0
Total transferred:      76291000 bytes
HTML transferred:       76127000 bytes
Requests per second:    11.83 [#/sec] (mean)
Time per request:       845.606 [ms] (mean)
Time per request:       84.561 [ms] (mean, across all concurrent requests)
Transfer rate:          881.06 [Kbytes/sec] received
```

Much better than master, but still worse than 0.8.1-RC1.

With reduced background usage (only `/v2/apps?embed=apps.tasks` at 3rps) CPU usage is roughly the same at 320ms, but max RPS differ:

0.8.1-RC1:

```
Document Path:          /v2/apps?embed=apps.tasks
Document Length:        72913 bytes

Concurrency Level:      10
Time taken for tests:   47.749 seconds
Complete requests:      1000
Failed requests:        0
Write errors:           0
Total transferred:      73077000 bytes
HTML transferred:       72913000 bytes
Requests per second:    20.94 [#/sec] (mean)
Time per request:       477.487 [ms] (mean)
Time per request:       47.749 [ms] (mean, across all concurrent requests)
Transfer rate:          1494.58 [Kbytes/sec] received
```

PR 1568:

```
Document Path:          /v2/apps?embed=apps.tasks
Document Length:        75623 bytes

Concurrency Level:      10
Time taken for tests:   70.917 seconds
Complete requests:      1000
Failed requests:        0
Write errors:           0
Total transferred:      75787000 bytes
HTML transferred:       75623000 bytes
Requests per second:    14.10 [#/sec] (mean)
Time per request:       709.175 [ms] (mean)
Time per request:       70.917 [ms] (mean, across all concurrent requests)
Transfer rate:          1043.62 [Kbytes/sec] received
```

Go ahead with RC, i'll reduce the load with label selectors, sse and probably mesos api as the source of truth.

**Comment by kolloch** [comment]
Hi @bobrik, the strange thing is that the metrics that you gave us earlier told a different story.

If you want, you can still send us the related metrics and I'll have a look.

We will not release master as an RC but the old RC with only this single fix. Maybe it works better, maybe not.
